server/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -179,6 +182,18 @@ public Store(
         ShardLock shardLock,
         OnClose onClose,
         boolean hasIndexSort
+    ) {
+        this(shardId, indexSettings, directory, shardLock, onClose, hasIndexSort, true);
+    }
+
+    public Store(
+        ShardId shardId,
+        IndexSettings indexSettings,
+        Directory directory,
+        ShardLock shardLock,
+        OnClose onClose,
+        boolean hasIndexSort,
+        boolean detectLeaks
```

Comment:
I don't like the need to be able to turn leak detection off, but there were a few places where we don't get access to the store to manage its lifecycle correctly.

Reply:
I think we should try avoid this flag. Browing through the changes, it is used for 
1. ShardSnapshotTaskRunnerTests
2. SourceOnlySnapshotRepository

You suggested a solution for the first usage. Is it possible to manage the lifecycle properly for the 2nd use case? It seems to me that we do have full access to the `tmpStore`? I haven't dug into it deepy so likely miss something.

Reply:
Yeah agreed, I think it is a smell. I can look a bit harder for alternatives.

I think with `tmpStore` we can probably use `ActionListener.runBefore(context, tempStore::close)` or similar. I'll do some digging to make sure that works.